### PR TITLE
sandbox: add a cors sandbox

### PR DESCRIPTION
**Pull request by dschaller**
*Description*: Add a sandbox for CORS filter.
*Risk Level*: Low
*Testing*: Manually tested the sandbox using the documented steps.
*Docs Changes*: Added documentation on the new sandbox.
*Release Notes*: Included
Fixes [#2167](https://github.com/envoyproxy/envoy/issues/2167)

### CODE REVIEW

File: examples/cors/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+To learn about this sandbox and for instructions on how to run it please head over
+to the [envoy docs](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/cors.html)
```
Comment (venilnoronha) [inline examples/cors/README.md]:
```suggestion
to the [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/cors.html).
```

File: examples/cors/frontend/service.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+app = Flask(__name__)
+
+
+@app.route('/')
+def index():
+    file_dir = os.path.dirname(os.path.realpath(__file__))
+    return send_from_directory(file_dir, 'index.html')
+
+
```
[the remark this answers is not in the document]
Comment (dschaller) [reply examples/cors/frontend/service.py]: I was going off of PEP8 which says there should be two carriage returns between functions. Envoy also has the same enforcement for the python tool scripts :)

[the remark this answers is not in the document]
Comment (venilnoronha) [reply examples/cors/frontend/service.py]: Oops, didn't realize that. Thanks.

**Comment by dschaller** [comment]
Thanks for the review @venilnoronha! I updated based on your feedback.

The only thing I wanted to ask about is the removal of `$` from the code blocks. I'm happy to remove it but it might be less clear what is input vs. output for some of the lines. What do you think?

**Comment by venilnoronha** [comment]
I think we can keep the `$` for all `console` code blocks in that case.

File: docs/root/start/sandboxes/cors.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  backend_backend-service_1_c7752ae7192a   /bin/sh -c /usr/local/bin/ ... Up      10000/tcp, 80/tcp
+
+**Step 4: Test Envoy's CORS capabilities**
+
+You can now open a browser to view your frontend service. To find the IP of
+your frontend service run the following command in terminal 1.
+
+.. code-block:: console
+
+  docker-machine ip frontend
```
Comment (venilnoronha) [inline docs/root/start/sandboxes/cors.rst]:
```suggestion
  $ docker-machine ip frontend
```

File: docs/root/start/sandboxes/cors.rst [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+**Step 6: Check stats of backend via admin**
+
+When Envoy runs, it can listen to ``admin`` requests if a port is configured. In the example
+configs, the admin is bound to port ``8001``.
+
+If you go to ``<backend_service_ip>:8001/stats`` you will be able to view
+all of the Envoy stats for the backend. You should see the CORS stats for
+invalid and valid origins increment as you make requests from the frontend cluster.
+
+.. code-block:: console
```
Comment (venilnoronha) [inline docs/root/start/sandboxes/cors.rst]:
```suggestion
.. code-block:: none
```

**Comment by dschaller** [comment]
Thanks @venilnoronha

File: docs/root/start/sandboxes/cors.rst [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+  $ docker-machine ip frontend
+
+In terminal 2, run the same command to get the remote origin IP that the
+frontend service will be making requests to.
+
+Results of the cross-origin request will be shown on the page under *Request Results*.
+Your browser's CORS enforcement logs can be found in the console.
```
Comment (venilnoronha) [inline docs/root/start/sandboxes/cors.rst]: One last thing. It'd be good to add some results and sample logs for the 3 cases here.

Comment (dschaller) [reply docs/root/start/sandboxes/cors.rst]: I've included the console logs for CORS errors. For the two failure cases the errors are the same and for the success case, there are no console errors.

File: docs/root/start/sandboxes/cors.rst [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+For example:
+
+.. code-block:: javascript
```
Comment (venilnoronha) [inline docs/root/start/sandboxes/cors.rst]:
```suggestion
.. code-block:: console
```

**Comment by dschaller** [comment]
Can someone rerun the failed test? Pretty sure it’s just flake.

File: docs/root/intro/version_history.rst [hunk 7]
```diff
@@ -55,6 +55,7 @@ Version history
   been fully decoded by Envoy.
 * router: added support for not retrying :ref:`rate limited requests<config_http_filters_router_x-envoy-ratelimited>`. Rate limit filter now sets the :ref:`x-envoy-ratelimited<config_http_filters_router_x-envoy-ratelimited>`
   header so the rate limited requests that may have been retried earlier will not be retried with this change.
+* sandbox: added :ref:`cors sandbox <install_sandboxes_cors>`.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: alpha order

File: examples/BUILD [hunk 8]
```diff
@@ -10,6 +10,10 @@ envoy_package()
 filegroup(
     name = "configs",
     srcs = [
+        "cors/backend/front-envoy.yaml",
```
Comment (mattklein123) [inline examples/BUILD]: I would have expected https://github.com/envoyproxy/envoy/blob/master/test/config_test/example_configs_test.cc#L22 to need to change with these added configs, which makes me wonder if something is broken? Can you take a look and make sure these new configs are getting tested?

Comment (dschaller) [reply examples/BUILD]: Turns out this isn't broken but it doesn't account for multiple config files with the same name.

Fixed in https://github.com/envoyproxy/envoy/pull/5220

**Pull request merged**